backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/codec/impl/HibernateSearchLuceneCodec.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.lowlevel.codec.impl;
+
+import org.hibernate.search.backend.lucene.document.model.impl.LuceneIndexField;
+import org.hibernate.search.backend.lucene.document.model.impl.LuceneIndexModel;
+import org.hibernate.search.backend.lucene.types.codec.impl.LuceneFieldCodec;
+import org.hibernate.search.backend.lucene.types.codec.impl.LuceneVectorFieldCodec;
+import org.hibernate.search.engine.backend.document.model.spi.IndexFieldFilter;
+
+import org.apache.lucene.codecs.Codec;
+import org.apache.lucene.codecs.FilterCodec;
+import org.apache.lucene.codecs.KnnVectorsFormat;
+import org.apache.lucene.codecs.lucene95.Lucene95Codec;
+import org.apache.lucene.codecs.perfield.PerFieldKnnVectorsFormat;
+
+public class HibernateSearchLuceneCodec extends FilterCodec {
+
+	private static final Codec DEFAULT_CODEC = new Lucene95Codec();
+
+	private final KnnVectorsFormat knnVectorsFormat;
+
+	public HibernateSearchLuceneCodec(LuceneIndexModel model) {
+		this( new IndexModelBasedPerFieldKnnVectorsFormat( model ) );
+	}
+
+	public HibernateSearchLuceneCodec(KnnVectorsFormat knnVectorsFormat) {
+		super( HibernateSearchLuceneCodec.class.getSimpleName(), DEFAULT_CODEC );
+		this.knnVectorsFormat = knnVectorsFormat;
+	}
+
+	public HibernateSearchLuceneCodec() {
+		this( (KnnVectorsFormat) null );
```

Comment:
Lucene is loading all these codecs through service loader and if we do not register one the whole thing fails to at the startup.. we also cannot use `Codec.getDefault()` as it can fall under race conditions depending on which codec gets loaded first...

> Please note: Don't call Codec.forName from the no-arg constructor of your own codec. When the SPI framework loads your own Codec as SPI component, SPI has not yet fully initialized! If you want to extend another Codec, instantiate it directly by calling its constructor.

that's why I've just used that constant for the codec and then added a test that will just fail if the `Codec.getDefault != Lucene95Codec` so that we'll notice if the thing will get updated...

Also I am not 100% convinced that this is a good approach, as this codec is still getting loaded and set in some lucene internals, but from some small experiments I've tried, it seemed to work ok...

Reply:
From what I understand, this loads the codec to read a segment that was written previously, and the name of the codec to load is included in the segment.

So, two things:

1. Yes, if it only uses the codec for the segment infos format, it probably doesn't need our codec wrapper.
2. We should make sure that the name of our own codec doesn't get written into the segment, because that would mean the segment would no longer include information about which "format" (Lucene version) it was written for. It would load our own codec, but delegating to the latest Lucene codec instead of the one it was written for... This just removes any chance of being able to read an index written with a previous version of Hibernate Search/Lucene.

I think #2 probably warrants a test, to be extra sure that the codec included in segment infos resolves to the Lucene one, not ours whose version is hard-coded...
We can access the codec from an `IndexReader` with `((StandardDirectoryReader) reader).getSegmentInfos().iterator().next().info`.
I don't know what would make more sense, a unit test in `backend/lucene/src/test/java/org/hibernate/search/backend/lucene/lowlevel/writer/impl` or an integration test in `integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/lowlevel/writer`. Whatever's simpler, I guess?

Reply:
added something for this in the latest commit; let me know what do you think about it 😃

Reply:
Thanks. The approach looks good to me, I added a few comments.